pkix/cert_auth.go

```diff
@@ -69,7 +78,7 @@ func CreateCertificateAuthority(key *Key, organizationalUnit string, expiry time
 
 // CreateIntermediateCertificateAuthority creates an intermediate
 // CA certificate signed by the given authority.
-func CreateIntermediateCertificateAuthority(crtAuth *Certificate, keyAuth *Key, csr *CertificateSigningRequest, proposedExpiry time.Time) (*Certificate, error) {
+func CreateIntermediateCertificateAuthority(crtAuth *Certificate, keyAuth *Key, csr *CertificateSigningRequest, proposedExpiry time.Time, pathlen int) (*Certificate, error) {
```

Comment:
Sorry, it's taken me so long to review these changes because I needed to be sure that this change to the public `pkix` package wouldn't break anyone. Unfortunately, looks like it will, for example https://cs.github.com/cloudfoundry/locket/blob/ad30c800960d38a0090acd7aa2acfc3f439b1f82/cmd/locket/certauthority/certauthority.go#L88.

Since we're not ready for a v2, I propose we add new functions (`CreateIntermediateCertificateAuthorityWithOptions` and `CreateCertificateAuthorityWithOptions`) to take `opts ...Option` to keep backwards compatibility: 

```
type Option func(*x509.Certificate)

func WithPathlen(pathlen int, excludePathlen bool) {
  return func(template *x509.Certificate) {
    // Set MaxPathLen appropriately....
  }
}
```

Reply:
(Note that part of the v2 changes will be moving things into an `internal` package so that it's clear we won't guarantee backward compatibility of thee APIs -- this isn't the first time it's hampered contributions. Thanks again for your patience.)

Reply:
Okay, let me revert the original function back. I kinda don't want other people code to break as well :D

Reply:
For the `init` command, it is okay to add the extra flag right?

Reply:
Would this work?

```go
pkix.CreateCertificateAuthorityWithOption(key, c.String("organizational-unit"), expiresTime, c.String("organization"), c.String("country"), c.String("province"), c.String("locality"), c.String("common-name"), c.StringSlice("permit-domain"), pkix.WithPathlenOption(c.Int("path-length"), c.Bool("exclude-path-length")))
```

Reply:
For the following flags should just leave it as before or create a func similar to `WithPathlen`?

Reply:
https://github.com/square/certstrap/blob/a22b32b990b4b05aa51678493a7187b21f276418/cmd/init.go#L193-L197

Reply:
Which approach should I take?

```go
 var opts = []pkix.Option{ 
 	pkix.WithPathlenOption(c.Int("pathlen"), c.Bool("exclude-path-length")), 
 } 
  
 crt, err := pkix.CreateCertificateAuthorityWithOption(key, c.String("organizational-unit"), expiresTime, c.String("organization"), c.String("country"), c.String("province"), c.String("locality"), c.String("common-name"), c.StringSlice("permit-domain"), opts) 
```

Or

```go
 crt, err := pkix.CreateCertificateAuthorityWithOption(key, c.String("organizational-unit"), expiresTime, c.String("organization"), c.String("country"), c.String("province"), c.String("locality"), c.String("common-name"), c.StringSlice("permit-domain"), pkix.WithPathlenOption(c.Int("pathlen"), c.Bool("exclude-path-length"))) 
```

Reply:
Yes, I think it's fine to have the exclude flag in `init`.